extensions/ql-vscode/src/remote-queries/remote-queries-monitor.ts

```diff
@@ -0,0 +1,97 @@
+import * as vscode from 'vscode';
+import { Credentials } from '../authentication';
+import { Logger } from '../logging';
+import { showAndLogErrorMessage } from '../helpers';
+import { RemoteQuery } from './remote-query';
+import { RemoteQueryWorkflowResult } from './remote-query-workflow-result';
+
+export class RemoteQueriesMonitor {
+  private static readonly twoDaysInSeconds = 86400;
+
+  constructor(
+    private readonly extensionContext: vscode.ExtensionContext,
+    private readonly logger: Logger
+  ) {
+  }
+
+  public async monitorQuery(
+    remoteQuery: RemoteQuery,
+    cancellationToken: vscode.CancellationToken
+  ): Promise<RemoteQueryWorkflowResult> {
+    const credentials = await Credentials.initialize(this.extensionContext);
+
+    if (!credentials) {
+      await showAndLogErrorMessage('Error authenticating with GitHub');
+    }
+
+    let attemptCount = 0;
+
+    const octokit = await credentials.getOctokit();
+
+    while (attemptCount <= RemoteQueriesMonitor.twoDaysInSeconds) {
```

Comment:
This expression is comparing the number of attempts with seconds. I see below that you are sleeping for a second between each call, but each call will be more than a second (since it's waiting for a network call to return). So, it's not really going to be two days in the end.

Also, if we decide to wait for more than 1s between calls, the name and value of the constant will need to change.

Suggestions:

1. Choose a max number of calls to the server, instead of a max time
2. Or, instead of sleeping for a fixed amount each time, choose the amount of time to sleep based on the current clock time. Eg- if we want to wait 5 seconds between calls, sleep for this amount:

    ```
    this.sleep(5000 - Date.now() + timeOfLastCall); // need to be careful if timeOfLastCall is more than 5 seconds ago.
    ```

And the while loop expression could look something like this:

```suggestion
    const startTime = Date.now();
    while ((Date.now() - startTime) / 1000 > RemoteQueriesMonitor.twoDaysInSeconds) { 
```

This will give a little more consistency with your checking and make it easier to change intervals.

Reply:
I've gone with the first suggestion which is what I had in mind in the first place but failed to tidy up the code - max number of attempts. I think it keeps the code simpler. And since we don't have a product requirement to check for X hours, it seems more straightforward to do that. Things might change when we start managing query history etc.